### PR TITLE
Accept supplementary characters

### PR DESCRIPTION
**Pull request by som-snytt**
Forward port https://github.com/scala/scala/pull/9687

Fixes https://github.com/lampepfl/dotty/issues/13045

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/Pickler.scala [hunk 1]
```diff
@@ -137,11 +137,14 @@ class Pickler extends Phase {
   }
 
   private def testSame(unpickled: String, previous: String, cls: ClassSymbol)(using Context) =
-    if (previous != unpickled) {
+    import java.nio.charset.StandardCharsets.UTF_8
+    def normal(s: String) = new String(s.getBytes(UTF_8), UTF_8)
+    val unequal = unpickled.length() != previous.length() || normal(unpickled) != normal(previous)
+    if unequal then
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/transform/Pickler.scala:143]: I didn't track down the encoding issue with unpickled as compilationUnit.toString. The behavior was that strings are unequals but array dump to txt files are the same.

**Comment by som-snytt** [comment]
Includes mild `getLitChar` refactor from scala 2 PR and adds unicode arrow deprecation.

**Review by dwijnand — APPROVED** [review]
LGTM, but it's not my area of expertise so I'll merge it in a few days to see if there are other reviewers.

**Comment by som-snytt** [comment]
Should nail down the unpickle test.

**Comment by som-snytt** [comment]
Rebased, but let's wait a few more days.

**Comment by odersky** [comment]
> LGTM, but it's not my area of expertise so I'll merge it in a few days to see if there are other reviewers.

It's been a few days 😉 Merge now?

**Comment by michelou** [comment]
@som-snytt Nice refactoring (`charEscape`, `uEscape`, `octalEscape`) ! 👍

**Pull request merged**